sql/v2/errors/errors.go

```diff
@@ -0,0 +1,151 @@
+/*
+ Copyright 2024 The CloudEvents Authors
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+package errors
+
+import (
+	"fmt"
+	"strings"
+)
+
+type errorKind int
+
+const (
+	parseError = iota
+	mathError
+	castError
+	missingAttributeError
+	missingFunctionError
+	functionEvaluationError
+)
+
+type cesqlError struct {
+	kind    errorKind
+	message string
+}
+
+func (eKind errorKind) String() string {
+	switch eKind {
+	case parseError:
+		return "parse error"
+	case mathError:
+		return "math error"
+	case castError:
+		return "cast error"
+	case missingAttributeError:
+		return "missing attribute error"
+	case missingFunctionError:
+		return "missing function error"
+	case functionEvaluationError:
+		return "function evaluation error"
+	default:
+		return "generic error"
+	}
+
+}
+
+func (cerr cesqlError) Error() string {
+	return fmt.Sprintf("%s: %s", cerr.kind, cerr.message)
+}
+
+func NewParseError(errs []error) error {
+	if len(errs) == 0 {
+		return nil
+	}
+
+	errorMessages := make([]string, 0, len(errs))
+	for _, err := range errs {
+		errorMessages = append(errorMessages, err.Error())
+	}
+
+	return cesqlError{
+		kind:    parseError,
+		message: strings.Join(errorMessages, ","),
+	}
+}
+
+func IsParseError(err error) bool {
+	if cesqlErr, ok := err.(cesqlError); ok {
+		return cesqlErr.kind == parseError
+	}
+	return false
+}
+
+func IsMathError(err error) bool {
+	if cesqlErr, ok := err.(cesqlError); ok {
+		return cesqlErr.kind == mathError
+	}
+	return false
+}
+
+func NewMathError(message string) error {
+	return cesqlError{
+		kind:    mathError,
+		message: message,
+	}
+}
+
+func IsCastError(err error) bool {
+	if cesqlErr, ok := err.(cesqlError); ok {
+		return cesqlErr.kind == castError
+	}
+	return false
+}
+
+func NewCastError(err error) error {
+	return cesqlError{
+		kind:    castError,
+		message: err.Error(),
+	}
+}
+
+func IsMissingAttributeError(err error) bool {
+	if cesqlErr, ok := err.(cesqlError); ok {
+		return cesqlErr.kind == missingAttributeError
+	}
+	return false
+}
+
+func NewMissingAttributeError(attribute string) error {
+	return cesqlError{
+		kind:    missingAttributeError,
+		message: attribute,
+	}
+}
+
+func IsMissingFunctionError(err error) bool {
+	if cesqlErr, ok := err.(cesqlError); ok {
+		return cesqlErr.kind == missingFunctionError
+	}
+	return false
+}
+
+func NewMissingFunctionError(function string) error {
+	return cesqlError{
+		kind:    missingFunctionError,
+		message: function,
+	}
+}
+
+func IsFunctionEvaluationError(err error) bool {
+	if cesqlErr, ok := err.(cesqlError); ok {
+		return cesqlErr.kind == functionEvaluationError
+	}
+	return false
+}
+
+func NewFunctionEvaluationError(err error) error {
+	return cesqlError{
+		kind:    functionEvaluationError,
+		message: err.Error(),
+	}
+}
+
+func IsGenericError(err error) bool {
+	if cesqlErr, ok := err.(cesqlError); ok {
+		return cesqlErr.kind < parseError || cesqlErr.kind > functionEvaluationError
```

Comment:
This feels risky to me. If someone adds a new error type will they remember to update this "return" line? I wonder if it would be safer to add a dummy error type to the end of the "const" list and then say `>= dummyLastError` here instead?

Reply:
I'd probably also use "0" instead of "parseError" in case someone adds a new item to the front of the const list.